New research concept: Uncertified robot modules.

Sunbeamstress opened the pull request:
🆑 
rscadd: With a bit of research into Data and Illegal technologies, the Roboticist can now print an uncertified robot module. What could it possibly do?
/🆑 

Uncertified modules are secret loadouts for borgs that can't be chosen from roundstart. They require research into illegal tech to unlock, at which point they can be produced from the Exosuit Fabricator. The intention is for them to be fun, benign, secret little modules, but there could be potential for combatative modules, or perhaps "upgraded" ones (a variant of the Engineering module with a plasteel synthesizer and RCD, for example).

These modules are special in that they are forcibly installed - using them on the robot will instantaneously change its module to the one specified on the board.

- **New module:** Party. An almost completely useless loadout featuring a boom box, strobe light, air horn, and money cannon. Rock out with it or kill on sight? Up to you!

https://www.youtube.com/watch?v=24T7y8Uzq0w&feature=youtu.be

- **New (old) module:** Combat. Oh yes, the one we all love to hate. It's been made into an uncertified module. But don't fear - it's still completely inaccessible by players.

Sunbeamstress committed:
Second squash!

Creates a mob-level proc for resetting robot modules, makes the various upgrade boards use that instead.

Removes unneeded uses of src.

no wait that was dumb

Additional tweaks.

Cleanup of qdel()-->null operations, fixing Initialize()s, and moving things around various procs, requested by PsiOmegaDelta

- SQUASH!!!
- Removed the hacky on-mob implementation of the strobe effect. The party light now properly creates and handles the movement of its own flashing light. Added bonus: badmins can give party lights to mobs other than robots if they want to.
- Made the air horn slightly less of an absolute unit.
- Added an uncertified module for the Combat robot. It remains an admin-only item, but should make it a lot easier to create one than the usual fucky method.
- Renamed the party light's state in lens_flare just to satisfy my own need for organization.

Changes requested by PsiOmegaDelta.

An \improper addition and also a module icon to appease almighty Travis.

Forgot to remove some commented out code.

Adds uncertified borg modules, and a party borg to start things off.

MistakeNot4892 commented on code/game/objects/items/devices/boombox.dm:
```diff
@@ -38,8 +38,8 @@
 
-/obj/item/device/boombox/attack_ai(mob/user)
-	return //silicons have no groove
+/obj/item/device/boombox/Topic(href, href_list, state = GLOB.physical_state)
```
This has no purpose. Can be removed.

MistakeNot4892 replied:
Discussion on Discord reveals this is needed for borgs to play boomboxes.

PsiOmegaDelta replied:
I'll be making it possible to define the default `state` in a more reasonable way.

MistakeNot4892 commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,45 @@
+	return 1
+
+/obj/item/borg/upgrade/uncertified/party
+	name = "Madhouse Productions Official Party Module"
```
`\improper`

afterthought2 commented:
Can we get a gif of the light effect? Icon diff bot is down or something.

I like the idea. I just want to make sure the strobing animation won't make me exit/minimize the game on sight (if it's animated).

EDIT: I'm dumb and I didn't see the link.

Sunbeamstress commented:
@afterthought2 check the youtube link at the bottom.

PsiOmegaDelta commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,45 @@
+
+/obj/item/borg/upgrade/uncertified/action(var/mob/living/silicon/robot/R)
+	if(..())
+		return 0
```
If true return false?

Sunbeamstress replied:
This is what I get for copy/pasting.

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -13,6 +13,7 @@
 	mob_push_flags = ~HEAVY //trundle trundle
 
 	var/lights_on = 0 // Is our integrated light on?
+	var/strobing = 0
```
Why is this item effect a mob level var?

Sunbeamstress replied:
I could not find any other way to get the strobe light to display an overlay over the borg.

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -248,21 +249,26 @@
 
-/mob/living/silicon/robot/proc/pick_module()
-	if(module)
+/mob/living/silicon/robot/proc/pick_module(var/override = 0)
```
`override` appears to be intended to be a module, so why is the default 0 instead of null?

Sunbeamstress replied:
lol, because I don't know what I'm doing

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -792,3 +793,23 @@ var/global/list/robot_modules = list(
+	sprites = list(	"Roller" = "omoikane"
+			  )
+
+/obj/item/weapon/robot_module/uncertified/party/New()
```
Use `/ Initialize()`
Return the result of `..()`

PurpleMartinJCK commented:
How will this work with custom robot skins?

Sunbeamstress commented:
In this current implementation, the one module ("party") has a single skin ("omoikane?"). It simply forces you into that skin.

MistakeNot4892 commented:
Custom borg skins override module icon selection. People will need to add a party module to their icon set.

Sunbeamstress commented:
@PsiOmegaDelta will the mob-level implementation be a deal breaker?

Sunbeamstress commented:
Thank you to @PsiOmegaDelta for his help!

- Removed the hacky on-mob implementation of the strobe effect. The party light now properly creates and handles the movement of its own flashing light. Added bonus: badmins can give party lights to mobs other than robots if they want to.
- Made the air horn slightly less of an absolute unit.
- Added an uncertified module for the Combat robot. It remains an admin-only item, but should make it a lot easier to create one than the usual fucky method.
- Renamed the party light's state in lens_flare just to satisfy my own need for organization.
- Squashed this silly boy for cleanliness (aka godliness).

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot_items.dm:
```diff
@@ -97,6 +97,70 @@
+	// Cause the party light effect to stop following this object, and then delete it.
+	GLOB.moved_event.unregister(src, src.strobe_effect, /atom/movable/proc/move_to_turf_or_null)
+	qdel(src.strobe_effect)
+	src.strobe_effect = null
```
`QDEL_NULL(strobe_effect)` does the above as a single line.

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot_items.dm:
```diff
@@ -97,6 +97,70 @@
+		activated = 1
+		set_light(1, 1, 7)
+		icon_state = "partylight-on"
+		item_state = "partylight-on"
```
Don't update icon/item state or light outside of `/update_icon()`, `/New()` or `/Initialize()`

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot_items.dm:
```diff
@@ -97,6 +97,70 @@
+	update_icon()
+
+/obj/item/weapon/party_light/proc/update_strobe()
+	if (!src.activated)
```
This var appears to be only strenuously connected to the actual action of activating/de-activating the strobe.
Seems like something best set by `activate_strobe()`/`deactivate_strobe()`

PsiOmegaDelta commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,52 @@
+	var/new_module = null
+
+/obj/item/borg/upgrade/uncertified/action(var/mob/living/silicon/robot/R)
+	if(..())
```
Should change this to `if ((. = ..()))`

PsiOmegaDelta commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,52 @@
+	desc = "A well-made but somewhat archaic looking bit of chipboard."
+	new_module = "Combat"
+
+/obj/item/borg/upgrade/uncertified/combat/Initialize()
```
Need to call `. = ..()`

PsiOmegaDelta commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,52 @@
+	desc = "A weird-looking chip with third-party additions crudely soldered in. It feels cheap and chintzy in the hand."
+	new_module = "Party"
+
+/obj/item/borg/upgrade/uncertified/party/Initialize()
```
Need to call `. = ..()`

PsiOmegaDelta commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,52 @@
+
+	if(R.module)
+		R.module.Reset(R)
+		qdel(R.module)
```
Another case of `QDEL_NULL()` wrapping multiple lines in one.
Why is `R.module = null` not intended anyway?

Sunbeamstress replied:
It has to do with the module's Reset() proc, it runtimes if there's no module specified - I did it like this because borgs created with the (admittedly less than ideal) spawn verb don't seem to initialize quite like ones created or roundstarted the normal way.

If I use QDEL_NULL(), it'll result in modules failing to work in any way whatsoever on the robot, as it can't reset.

PsiOmegaDelta replied:
What does using QDEL_NULL() instead of qdel+null have to do with resetting?

PsiOmegaDelta commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -37,6 +37,52 @@
+	R.uneq_all()
+	R.modtype = initial(R.modtype)
+
+	R.hands.icon_state = initial(R.hands.icon_state)
```
This is iffy. Why is this needed exactly?

Sunbeamstress replied:
Copy/paste from /obj/item/borg/upgrade/reset. Falls under the case of "I don't like it either but the entire universe is set on fire if I remove it"

PsiOmegaDelta commented on code/modules/research/designs.dm:
```diff
@@ -2217,3 +2217,13 @@ CIRCUITS BELOW
 	sort_string = "XACAD"
+
+/datum/design/item/party
+	name = "uncertified module: PRTY"
```
Capitalize "uncertified"

PsiOmegaDelta commented:
What's with the use of `src` all over the place?

Sunbeamstress commented:
@PsiOmegaDelta tbh my use of `src` came from hearing Zuhayr rant about relative pathing, I was led to believe it was a safer option

dryerlint commented:
Which music track is the one playing in the Youtube video?

MistakeNot4892 commented:
@Earthcrusher for reference relative pathing vs absolute pathing is
````
obj
	item
		proc
			foo()
				return "foo"
````
vs
````
/obj/item/proc/foo()
	return "foo"
````
Use of `src` is unrelated.

MistakeNot4892 commented:
that was not the comment button

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot_items.dm:
```diff
@@ -97,6 +97,67 @@
+
+/obj/item/weapon/party_light/proc/activate_strobe()
+	activated = 1
+	set_light(1, 1, 7)
```
Didn't I just say : Do not set light outside of update_icon()?

Sunbeamstress replied:
You did, I missed it. I appreciate your help but can you chill some?

Sunbeamstress commented:
I think maybe I'm unclear about the boundaries between "I have a curious question about your implementation" and "you need to do this for me to approve it" so I'm just gonna remove redundant uses of `src` and see how you guys like it

Sunbeamstress committed:
fixed the formatting of Initialize() procs

Sunbeamstress commented:
Moved the redundant functionality from the reset and uncertified modules and made them into a mob-level proc on the robot itself.

Added bonus: I've confirmed with testing that it's now safe to use upgrade modules on adminspawned robots!



Edit: Also altered the OP a bit, since this increased in scope

PsiOmegaDelta commented on code/modules/mob/living/silicon/robot/robot_items.dm:
```diff
@@ -97,6 +97,67 @@
+	else
+		activate_strobe()
+
+	update_icon()
```
If the strobe procs are the ones changing the appearance they should be the ones calling `update_icon()`

Sunbeamstress committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into UncertifiedBorgModules

MistakeNot4892 reviewed (dismissed):
I don't see anything too remiss.

Sunbeamstress committed:
Shuffles around some update_icons(), fixes a few descs.

Earthcrusher commented:
Just FYI there's another commit that you can see here at https://github.com/Baystation12/Baystation12/compare/dev...Earthcrusher:UncertifiedBorgModules?expand=1

I have no idea why it's not showing up, other than Github recently shitting itself.
